cedar-policy-cli/src/lib.rs

```diff
@@ -1177,36 +1177,66 @@ fn read_human_policy_set(
     let json_source = read_from_file_or_stdin(filename, context)?;
-    let json: serde_json::Value = serde_json::from_str(&json_source).into_diagnostic()?;
-    let err_to_report = |err| {
+    let json = serde_json::from_str::<serde_json::Value>(&json_source).into_diagnostic()?;
+    let policy_type = get_json_policy_type(&json)?;
+
+    let add_json_source = |report: Report| {
         let name = filename.map_or_else(
             || "<stdin>".to_owned(),
             |n| n.as_ref().display().to_string(),
         );
-        Report::new(err).with_source_code(NamedSource::new(name, json_source.clone()))
+        report.with_source_code(NamedSource::new(name, json_source.clone()))
     };
 
-    match Policy::from_json(None, json.clone()) {
-        Ok(policy) => PolicySet::from_policies([policy])
-            .wrap_err_with(|| format!("failed to create policy set from {context}")),
-        Err(_) => match Template::from_json(None, json).map_err(err_to_report) {
-            Ok(template) => {
-                let mut ps = PolicySet::new();
-                ps.add_template(template)?;
-                Ok(ps)
-            }
-            Err(err) => Err(err).wrap_err_with(|| format!("failed to parse {context}")),
+    match policy_type {
+        JsonPolicyType::SinglePolicy => match Policy::from_json(None, json.clone()) {
+            Ok(policy) => PolicySet::from_policies([policy])
+                .wrap_err_with(|| format!("failed to create policy set from {context}")),
+            Err(_) => match Template::from_json(None, json)
+                .map_err(|err| add_json_source(Report::new(err)))
+            {
+                Ok(template) => {
+                    let mut ps = PolicySet::new();
+                    ps.add_template(template)?;
+                    Ok(ps)
+                }
+                Err(err) => Err(err).wrap_err_with(|| format!("failed to parse {context}")),
+            },
         },
+        JsonPolicyType::PolicySet => {
+            PolicySet::from_json_value(json).map_err(|err| add_json_source(Report::new(err)))
+        }
     }
 }
 
+fn get_json_policy_type(json: &serde_json::Value) -> Result<JsonPolicyType> {
+    let policy_set_properties = ["staticPolicies", "templates", "templateLinks"];
+    let policy_properties = ["action", "effect", "principal", "resource", "conditions"];
+
+    let json_has_property = |p| json.get(p).is_some();
+    let has_any_policy_set_property = policy_set_properties.iter().any(json_has_property);
+    let has_any_policy_property = policy_properties.iter().any(json_has_property);
+
+    match (has_any_policy_set_property, has_any_policy_property) {
+        (false, false) => Err(miette!("Cannot determine if json policy is a single policy or a policy set. Found no matching properties from any of those schemas.")),
+        (true, true) => Err(miette!("Cannot determine if json policy is a single policy or a policy set. Found matching properties from both schemas.")),
```

Comment:
Adjust capitalization for consistency with other Cedar error messages
```suggestion
        (false, false) => Err(miette!("cannot determine if json policy is a single policy or a policy set. Found no matching properties from either format")),
        (true, true) => Err(miette!("cannot determine if json policy is a single policy or a policy set. Found matching properties from both formats")),
```